### PR TITLE
Revise SetHead method before applying the snapshot package

### PR DESCRIPTION
**Pull request by hqjang-pepper**
## Proposed changes

- This is the first revision of `SetHead` method, which should be applied before snapshot package is merged.
- `SetHead` method rewinds the chain to the given number, including currentBlock, currentFastBlock and currentHeader.
- If a target block doesn't have state information, the currentBlock is further rewinded to the block in which state is stored.
- You can rewind node by calling `debug_setHead` api.
- Added a flag called `tmpStop` to prevent synchronization temporarily while setHead is in progress.
- When the snapshot package is merged, it is necessary to apply it to setHead.

For PR reviewers, these are the two main changes of setHead from this PR
- For loop repeats until it finds 1. a block with a state 2. that is less than the target number.
- While rewinding a chain, downloader starts and get a chain segment from a peer. After block rewind successfully ends, node tries to add the chain segments to `rewinded chain`, which causes sync error. That's the reason why `tmpStop` is needed. 

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by jeongkyun-oh** [comment]
@hqjang-pepper I guess this is not referenced from the recent Ethereum codes. Could you add a PR link which you referenced?

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -450,54 +450,66 @@ func (bc *BlockChain) loadLastState() error {
...
-// above the new head will be deleted and the new one set. In the case of blocks
-// though, the head may be further rewound if block bodies are missing (non-archive
-// nodes after a fast sync).
 func (bc *BlockChain) SetHead(head uint64) error {
-	logger.Info("Rewinding blockchain", "target", head)
+	logger.Warn("Rewinding blockchain", "target", head)
 
-	bc.mu.Lock()
-	defer bc.mu.Unlock()
+	bc.chainmu.Lock()
```
Comment (jeongkyun-oh) [inline blockchain/blockchain.go:456]: mutex is changed, is this okay?

File: storage/database/db_manager.go [hunk 2]
```diff
@@ -823,6 +824,22 @@ func (dbm *databaseManager) DeleteCanonicalHash(number uint64) {
...
 }
 
+// ReadAllHashes retrieves all the hashes assigned to blocks at a certain heights,
+// both canonical and reorged forks included.
+func (dbm *databaseManager) ReadAllHashes(number uint64) []common.Hash {
+	db := dbm.getDatabase(headerDB)
+	prefix := headerKeyPrefix(number)
+
+	hashes := make([]common.Hash, 0, 1)
+	it := db.NewIterator(prefix, nil)
```
Comment (jeongkyun-oh) [inline storage/database/db_manager.go:834]: Don't we need `it.Release` after using it?

File: storage/database/db_manager.go [hunk 3]
```diff
@@ -823,6 +824,22 @@ func (dbm *databaseManager) DeleteCanonicalHash(number uint64) {
 	dbm.cm.writeCanonicalHashCache(number, common.Hash{})
 }
 
+// ReadAllHashes retrieves all the hashes assigned to blocks at a certain heights,
+// both canonical and reorged forks included.
+func (dbm *databaseManager) ReadAllHashes(number uint64) []common.Hash {
```
Comment (jeongkyun-oh) [inline storage/database/db_manager.go:829]: Where is this method used?

**Comment by hqjang-pepper** [comment]
> @hqjang-pepper I guess this is not referenced from the recent Ethereum codes. Could you add a PR link which you referenced?

- SetHead of blockchain.go - https://github.com/ethereum/go-ethereum/commit/542df8898ef6d718647058c129069804bc463ea5
- SetHead of headerchain.go - https://github.com/ethereum/go-ethereum/commit/b02958b9c57f35d0df085ed8e7057a52131373a3
- https://github.com/ethereum/go-ethereum/commit/8cbdc8638fd28693f84d7bdbbdd587e8c57f6383

File: blockchain/blockchain.go (hunk 1 above)
Comment (hqjang-pepper) [reply blockchain/blockchain.go:456]: Ethereum also changed mutex at https://github.com/ethereum/go-ethereum/commit/f25f776c9f9d6a0538cd9028d9d82fc6965d9123. Take a look.

File: storage/database/db_manager.go (hunk 2 above)
Comment (hqjang-pepper) [reply storage/database/db_manager.go:834]: I reflected. Thanks for reviewing.

File: storage/database/db_manager.go (hunk 3 above)
Comment (hqjang-pepper) [reply storage/database/db_manager.go:829]: @jeongkyun-oh 
Oh, this method will be used in `headerchain.go` **after** snapshot package is ported.

File: blockchain/blockchain.go (hunk 1 above)
Comment (aidan-kwon) [reply blockchain/blockchain.go:456]: @hqjang-pepper Your reference combined two mutexes into `bc.chainmu`. But your suggestion changes only one usage of the `bc.mu` mutex. It means that `SetHead` is not a critical section of `bc.mu` mutex and other codes still using `bc.mu` can be executed concurrently with `SetHead`. 
I think this mutex change is not proper for the intend of this PR.

Comment (hqjang-pepper) [reply blockchain/blockchain.go]: @aidan-kwon I reflected. Thanks for your comment.

File: node/cn/handler.go [hunk 4]
```diff
@@ -122,6 +122,8 @@ type ProtocolManager struct {
 
 	nodetype          common.ConnType
 	txResendUseLegacy bool
+
+	tmpstop bool
```
Comment (yoomee1313) [inline node/cn/handler.go:126]: The `tmpstop` is a little bit confusing to me to figure out what it does. What about changing to the more specific name?

File: node/cn/gen_config.go [hunk 5]
```diff
@@ -25,14 +25,19 @@ func (c Config) MarshalTOML() (interface{}, error) {
 		NetworkId               uint64
 		SyncMode                downloader.SyncMode
 		NoPruning               bool
+		WorkerDisable           bool
```
Comment (yoomee1313) [inline node/cn/gen_config.go:28]: Can I ask why these flags are added?

File: node/cn/api_backend.go [hunk 6]
```diff
@@ -75,8 +75,10 @@ func (b *CNAPIBackend) CurrentBlock() *types.Block {
 }
 
 func (b *CNAPIBackend) SetHead(number uint64) {
-	//b.cn.protocolManager.downloader.Cancel()
+	b.cn.protocolManager.Downloader().Cancel()
+	b.cn.protocolManager.SetTmpStop(true)
 	b.cn.blockchain.SetHead(number)
+	b.cn.protocolManager.SetTmpStop(false)
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply node/cn/api_backend.go:81]: 1. There is no way to call CN's protocolmanager in methods of `blockchain` package.
2. blockchain.SetHead is called only in two functions except API call. 
-  `NewBlockChain` is an initialize function so we don't need to care about peer sync.
- `ResetWithGenesisBlock`is called only in `loadLastState()` when Head block is missing or got some problem with db. And `loadLastState()` is used when making a new blockchain or reset the canonical blocks. Kind of initializing process. 

so I think in case of non-api function calls of `setHead`, we don't need to care about sync stop process.

File: node/cn/handler.go (hunk 4 above)
Comment (hqjang-pepper) [reply node/cn/handler.go:126]: Thx. I'll reflect.

File: blockchain/blockchain.go [hunk 7]
```diff
@@ -450,54 +450,90 @@ func (bc *BlockChain) loadLastState() error {
 	return nil
 }
 
-// SetHead rewinds the local chain to a new head. In the case of headers, everything
-// above the new head will be deleted and the new one set. In the case of blocks
-// though, the head may be further rewound if block bodies are missing (non-archive
-// nodes after a fast sync).
 func (bc *BlockChain) SetHead(head uint64) error {
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply blockchain/blockchain.go:453]: There is no **specific** reference version of this function.
Because necessary features are spread over several versions, and function format is different with Ethereum.
But mostly from these following PR --> https://github.com/ethereum/go-ethereum/commit/8cbdc8638fd28693f84d7bdbbdd587e8c57f6383

+ I'll add some comments about setHead. Don't remember why I just deleted. Thanks for reminding.

**Comment by yoomee1313** [comment]
@hqjang-pepper If you haven't, please test the setHead with istanbulCompatibleBlock set to larger than 0.

**Comment by ehnuje** [comment]
@hqjang-pepper This change got merge conflicts. Could you please resolve them and update the PR?

**Comment by hqjang-pepper** [comment]
> @hqjang-pepper This change got merge conflicts. Could you please resolve them and update the PR?

resolved.

File: api/mocks/backend_mock.go [hunk 8]
```diff
@@ -162,6 +162,7 @@ func (mr *MockBackendMockRecorder) Engine() *gomock.Call {
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "Engine", reflect.TypeOf((*MockBackend)(nil).Engine))
 }
 
+
```
Comment (kjhman21) [inline api/mocks/backend_mock.go:165]: unnecessary change.

File: node/cn/handler.go [hunk 9]
```diff
@@ -367,6 +370,11 @@ func (pm *ProtocolManager) Stop() {
 	logger.Info("Klaytn protocol stopped")
 }
 
+// SetSyncStop sets value of syncStop flag. If it's true, peer sync process does not proceed.
+func (pm *ProtocolManager) SetSyncStop(flag bool) {
+	pm.syncStop = flag
```
Comment (kjhman21) [inline node/cn/handler.go:375]: It seems that writing this variable and reading it can be done by different thread. In this case we need to use an atomic operation.

File: node/cn/sync.go [hunk 10]
```diff
@@ -182,8 +182,8 @@ func (pm *ProtocolManager) getSyncMode(currentBlock *types.Block) downloader.Syn
 
 // synchronise tries to sync up our local block chain with a remote peer.
 func (pm *ProtocolManager) synchronise(peer Peer) {
-	// Short circuit if no peers are available
-	if peer == nil {
+	// Short circuit if no peers are available or syncStop flag is set to true
+	if peer == nil || pm.syncStop {
```
Comment (kjhman21) [inline node/cn/sync.go:186]: same here. Don't we need to use atomic operation?

File: storage/database/db_manager.go [hunk 11]
```diff
@@ -823,6 +824,24 @@ func (dbm *databaseManager) DeleteCanonicalHash(number uint64) {
...
+	db := dbm.getDatabase(headerDB)
+	prefix := headerKeyPrefix(number)
+
+	hashes := make([]common.Hash, 0, 1)
```
Comment (kjhman21) [inline storage/database/db_manager.go:833]: Why do you make an array of hashes with capacity 1?

File: storage/database/schema.go [hunk 12]
```diff
@@ -105,6 +105,18 @@ type TxLookupEntry struct {
...
+// encodeBlockNumber encodes a block number as big endian uint64
+func encodeBlockNumber(number uint64) []byte {
+	enc := make([]byte, 8)
+	binary.BigEndian.PutUint64(enc, number)
+	return enc
+}
+
+// headerKeyPrefix = headerPrefix + num (uint64 big endian)
+func headerKeyPrefix(number uint64) []byte {
+	return append(headerPrefix, encodeBlockNumber(number)...)
```
Comment (kjhman21) [inline storage/database/schema.go:117]: Why don't you use `common.Int64ToByteBigEndian(number)` as shown below?

File: blockchain/blockchain_sethead_test.go [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+	}
+
+	// Set the head of the chain back to the requested number
+	chain.SetHead(tt.setheadBlock)
+
+	// Iterate over all the remaining blocks and ensure there are no gaps
+	//verifyNoGaps(t, chain, true, canonblocks)
+	//verifyNoGaps(t, chain, false, sideblocks)
+	//verifyCutoff(t, chain, true, canonblocks, tt.expCanonicalBlocks)
+	//verifyCutoff(t, chain, false, sideblocks, tt.expSidechainBlocks)
```
Comment (aidan-kwon) [inline blockchain/blockchain_sethead_test.go:160]: Why did you comment out these?

File: node/cn/api_backend.go [hunk 14]
```diff
@@ -76,8 +76,10 @@ func (b *CNAPIBackend) CurrentBlock() *types.Block {
 }
 
 func (b *CNAPIBackend) SetHead(number uint64) {
-	//b.cn.protocolManager.downloader.Cancel()
+	b.cn.protocolManager.Downloader().Cancel()
```
Comment (aidan-kwon) [inline node/cn/api_backend.go:79]: How downloader start again after `SetHead`?

**Comment by aidan-kwon** [comment]
`SetHead` removes old blocks but doesn't governance data. I think it should remove the governance data also because it can be stored in `misc` DB and affect the node later. @hqjang-pepper  Did you consider governance data also?

**Comment by kjhman21** [comment]
related to #1144

**Comment by hqjang-pepper** [comment]
> `SetHead` removes old blocks but doesn't governance data. I think it should remove the governance data also because it can be stored in `misc` DB and affect the node later. @hqjang-pepper Did you consider governance data also?

The fundamental question. Is clearing the governance DB data needed?
Even if `setHead()` method does not delete the governance DB data, I think it's okay.
While catching up the block, if there is governanceData that has already been entered in the governance DB, it will be fine because there is a logic to return without writing itself at all.

In function `WriteGovernance()`
```
if err := dbm.WriteGovernanceIdx(num); err != nil {
   if err == errGovIdxAlreadyExist {
      // Overwriting existing data is not allowed, but the attempt is not considered as a failure.
      return nil
   }
   return err
}
```

How do you think? @aidan-kwon @jeongkyun-oh

File: node/cn/handler.go (hunk 9 above)
Comment (hqjang-pepper) [reply node/cn/handler.go]: changed SetSyncStop method atomically.

File: node/cn/sync.go (hunk 10 above)
Comment (hqjang-pepper) [reply node/cn/sync.go]: Changed atomically, too.

**Comment by jeongkyun-oh** [comment]
> > `SetHead` removes old blocks but doesn't governance data. I think it should remove the governance data also because it can be stored in `misc` DB and affect the node later. @hqjang-pepper Did you consider governance data also?
> 
> The fundamental question. Is clearing the governance DB data needed? Even if `setHead()` method does not delete the governance DB data, I think it's okay. While catching up the block, if there is governanceData that has already been entered in the governance DB, it will be fine because there is a logic to return without writing itself at all.
> 
> In function `WriteGovernance()`
> 
> ```
> if err := dbm.WriteGovernanceIdx(num); err != nil {
>    if err == errGovIdxAlreadyExist {
>       // Overwriting existing data is not allowed, but the attempt is not considered as a failure.
>       return nil
>    }
>    return err
> }
> ```
> 
> How do you think? @aidan-kwon @jeongkyun-oh

As we talked, `setHead` may be called under the unexpected circumstance. If corrupted data may be (somehow) inserted and want to `setHead`, I think we should be able to insert new data. However, if wrong data may be inserted and we want to retry to recover governance data, it will not be since the logic you mentioned.

**Review by aidan-kwon — APPROVED** [review]

File: blockchain/blockchain.go [hunk 15]
```diff
@@ -450,53 +450,93 @@ func (bc *BlockChain) loadLastState() error {
...
+// that the rewind must pass the specified state root. The method will try to
+// delete minimal data from disk whilst retaining chain consistency.
 func (bc *BlockChain) SetHead(head uint64) error {
-	logger.Info("Rewinding blockchain", "target", head)
+	logger.Warn("Rewinding blockchain", "target", head)
 
 	bc.mu.Lock()
 	defer bc.mu.Unlock()
+	bc.chainmu.Lock()
+	defer bc.chainmu.Unlock()
```
Comment (jeongkyun-oh) [inline blockchain/blockchain.go:462]: Should we update this after https://github.com/klaytn/klaytn/pull/1099 ?

**Review by kjhman21 — APPROVED** [review]

File: blockchain/blockchain.go (hunk 15 above)
Comment (hqjang-pepper) [reply blockchain/blockchain.go]: I applied this change by rebasing.

File: node/cn/api_backend.go (hunk 14 above)
Comment (hqjang-pepper) [reply node/cn/api_backend.go:79]: After setHead logic is done, Td gets smaller than peer's. 
If local Td is smaller than peer's Td, then new goroutine starts downloader automatically.
```
if trueTD.Cmp(pm.blockchain.GetTd(currentBlock.Hash(), currentBlock.NumberU64())) > 0 {
	go pm.synchronise(p)
}
```

File: blockchain/blockchain_sethead_test.go (hunk 13 above)
Comment (hqjang-pepper) [reply blockchain/blockchain_sethead_test.go]: I commented this because of golint complexity.
PR https://github.com/klaytn/klaytn/pull/1137 is merged, so I uncommented.

File: storage/database/schema.go (hunk 12 above)
Comment (hqjang-pepper) [reply storage/database/schema.go]: I'll reflect this change. Thanks.

**Review by aidan-kwon — APPROVED** [review]

**Review by jeongkyun-oh — APPROVED** [review]

**Review by kjhman21 — APPROVED** [review]
